### DISCUSSION TITLE
Are atomic multi-repositories pushes possible?

### DISCUSSION THREAD

**Discussion by xmo-odoo** [community/community#120704] [Repositories] [answered]
### Select Topic Area

Question

### Body

With multi-repo / poly-repo projects, sometimes it's important for repositories to be updated in lockstep (especially when the project is badly managed or only multi-repo because of hype).

I know that git supports atomically updating multiple refs since 2.4, and [github supports that as well](https://github.blog/2015-04-30-git-2-4-atomic-pushes-push-to-deploy-and-more/), however is there a good way to perform atomic multi-repo updates?

I don't believe github provides repo locks, so an application-side 2PL or similar does not seem possible, and in my experience Github's cross-repository anything are rather limited.

**Comment by david-wei-01001** [comment, marked as answer]
As of my last update, there isn't a built-in mechanism in Git or GitHub to perform atomic transactions across multiple repositories. Git’s atomic operations generally apply to actions within a single repository, such as updating multiple refs in one push to that repository.

When dealing with multiple repositories (multi-repo or poly-repo), atomicity becomes more complex because each repository has its own commit history and ref namespace. Here are some considerations and potential workarounds:

1. **Submodules/Subtrees**: If you want to keep repositories in sync, you could consider using Git submodules or subtrees, which allow one repository to include content from another. However, this doesn't guarantee atomicity across repositories; it just simplifies the management of dependencies.

2. **CI/CD Pipeline**: Implement a CI/CD pipeline that updates each repository in sequence. While not atomic, this can ensure that if one step fails, the subsequent steps are not executed. Some CI/CD systems allow you to set up a pipeline across multiple repositories.

3. **Orchestration Tool**: Use an orchestration tool or script that sequentially updates each repository and can roll back changes if any step fails. While rollbacks can mitigate some of the risks, they are not equivalent to true atomic transactions.

4. **Mono-repo Approach**: Convert the multi-repo project into a mono-repo. This way, all related projects are part of the same repository, and atomic commits and rollbacks are naturally supported.

5. **Database Transactions**: For changes that must be applied across services and databases in a transactional way, you might need to implement a distributed transaction system with commit and rollback capabilities within your application logic.

6. **Manual Coordination**: For critical updates, manually coordinate and perform the updates in quick succession, possibly during a period of low activity. This requires strong communication and is error-prone but is sometimes used in practice.

7. **Tooling**: Use or build tooling that automates the above process as much as possible to reduce human error. Some teams develop custom scripts that trigger updates across all necessary repositories and include checks to ensure each update is successful before proceeding to the next.

In a cloud environment like GitHub, you generally do not have control over the repository locks that would be necessary to implement a two-phase locking protocol (2PL). This limits the options for true atomicity in multi-repo setups.

If atomic multi-repo updates are essential for your workflow, it may be worth reconsidering the multi-repo approach or implementing a strategy that minimizes the need for cross-repository consistency (e.g., decoupling services, using versioned contracts, etc.).

**Reply by xmo-odoo** [reply]
Thanks, that's about what I'd figured but I appreciate the exhaustivity. Sadly the setup is out of my hands (and furthermore generally has justifiable business reasoning behind it or at least some of it).

We do implement (2), (6), and (7) and the question was in part in that context: they're not foolproof / error-free, and as github is not always reliable these days and it can be difficult to envision all possible failure cases (or recover from them) (2) can have pretty difficult edge cases, and I wanted to know if there were ways to improve that I had missed. Sadly it seems that's not the case.

**Reply by Akash1134** [reply]
Hi @david-wei-01001 , 

We’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.
